modules/extensions-support/src/main/java/org/mule/module/extension/internal/manager/ExtensionRegister.java

```diff
@@ -0,0 +1,241 @@
+     */
+    ExtensionStateTracker getExtensionState(Extension extension)
+    {
+        return extensionStates.getUnchecked(extension);
+    }
+
+    /**
+     * @param configuration a {@link Configuration} owned by a registered {@link Extension}
+     * @return the {@link ExtensionStateTracker} object related to the given {@code configuration}
+     */
+    ExtensionStateTracker getExtensionState(Configuration configuration)
+    {
+        return getExtensionState(getExtension(configuration));
+    }
+
+    Map<String, ConfigurationInstanceProvider> getConfigurationInstanceProviders()
+    {
+        ImmutableMap.Builder<String, ConfigurationInstanceProvider> providers = ImmutableMap.builder();
+        for (Extension extension : extensions.values())
+        {
+            providers.putAll(getConfigurationInstanceProviders(extension));
+        }
+
+        return providers.build();
+    }
+
+    Map<String, ConfigurationInstanceProvider> getConfigurationInstanceProviders(Extension extension)
+    {
+        ImmutableMap.Builder<String, ConfigurationInstanceProvider> providers = ImmutableMap.builder();
+        for (Configuration configuration : extension.getConfigurations())
+        {
+            providers.putAll(getExtensionState(configuration).getConfigurationInstanceProviders(configuration));
+        }
+
+        return providers.build();
+    }
+
+    /**
+     * @param operation a {@link Operation} owned by a registered {@link Extension}
+     * @return the {@link ExtensionStateTracker} object related to the given {@code operation}
+     */
+    ExtensionStateTracker getExtensionState(Operation operation)
+    {
+        return getExtensionState(getExtension(operation));
+    }
+
+    /**
+     * @param capabilityType the {@link Class} of a capability
+     * @param <C>            the capability type
+     * @return an immutable view of all registered {@link Extension} which have the given {@code capabilityType}
+     */
+    <C> Set<Extension> getExtensionsCapableOf(Class<C> capabilityType)
+    {
+        Set<Extension> cachedCapables = capability2ExtensionCache.get(capabilityType);
+        if (CollectionUtils.isEmpty(cachedCapables))
+        {
+            ImmutableSet.Builder<Extension> capables = ImmutableSet.builder();
+            for (Extension extension : getExtensions())
+            {
+                if (extension.isCapableOf(capabilityType))
+                {
+                    capables.add(extension);
+                }
+            }
+
+            cachedCapables = capables.build();
+            capability2ExtensionCache.put(capabilityType, cachedCapables);
+        }
+
+        return cachedCapables;
+    }
+
+    private void clearCaches()
+    {
+        configuration2ExtensionCache.clear();
+        operation2ExtensionCache.clear();
+        capability2ExtensionCache.clear();
+    }
+
+    private <K, V> V lookupInCache(Map<K, V> cache, final K key, Predicate predicate)
```

Comment:
it could... but it's not been used anywhere else so I'd rather leave it here than to put it into a utility class that nobody knows about. If the need to reuse it ever comes, I'll be happy to move it